### PR TITLE
Homework_4

### PR DESCRIPTION
**Pull request by shotvlad**
Homework_4

### CODE REVIEW

File: Task_1/Task_1.cpp [hunk 1]
```diff
@@ -1,6 +1,118 @@
 ﻿#include <iostream>
 
+using namespace std;
+
+//ЧИСЛО СУДЬБЫ. Цифры полной даты рождения суммируются, пока не получится одна цифра. Разработайте функцию и вычислите свое число судьбы)
+
+int enter(int number, string str);
+bool check(int y, int m, int d);
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp]: `isCheck`

File: Task_1/Task_1.cpp [hunk 2]
```diff
@@ -1,6 +1,118 @@
...
 {
-    std::cout << "Hello World!\n";
+    int year = 0, month = 0, day = 0;
+
+    cout << "Enter your birthday: " << endl;
+    year = enter(year, "Year");
+    month = enter(month, "Month");
+    day = enter(day, "Day");
+
+    if (check(year, month, day) != true)
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp]: Refactoring `check(year, month, day) != true`

File: Task_1/Task_1.cpp [hunk 3]
```diff
@@ -1,6 +1,118 @@
...
+    {
+        cout << "\nThere is no such date." << endl;
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp:22]: without `\n`

File: Task_1/Task_1.cpp [hunk 4]
```diff
@@ -1,6 +1,118 @@
...
+
+        return -1;
+    }
+    else {}
+
+    cout << "\nNumber of fate: " << numberOfFate(year, month, day) << endl;
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp:28]: the same

File: Task_1/Task_1.cpp [hunk 5]
```diff
@@ -1,6 +1,118 @@
...
+    else if (m == 11 and (d < 1 or d > 30))
+    {
+        return false;
+    }
+    else if (m == 12 and (d < 1 or d > 31))
+    {
+        return false;
+    }
+    else return true;
 }
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp]: Bad function. Refactoring!

File: Task_2/Task_2.cpp [hunk 6]
```diff
@@ -1,6 +1,66 @@
...
+    if (number >= 0)
+    {
+        inverted = flipPositive(number);
+    }
+    else
+    {
+        inverted = flipNegative(number);
+    }
+
+    cout << "\nInverted number: " << inverted << endl;
```
Comment (VadzimAlkhimionak) [inline Task_2/Task_2.cpp]: without \n

File: Task_2/Task_2.cpp [hunk 7]
```diff
@@ -1,6 +1,66 @@
...
+    int flip = 0;
+
+    while (number < 0)
+    {
+        flip = (flip * 10) + (number % 10);
+
+        number /= 10;
+    }
+
+    return flip;
```
Comment (VadzimAlkhimionak) [inline Task_2/Task_2.cpp]: Why?
This is the same functions

Comment (VadzimAlkhimionak) [reply Task_2/Task_2.cpp]: Refactoring

**Review by VadzimAlkhimionak — COMMENTED** [review]
Hi. Comments for you. Don't repeat yourself. Read about it.

File: Task_1/Task_1.cpp (hunk 4 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp:28]: image

File: Task_1/Task_1.cpp (hunk 3 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp:22]: image

File: Task_2/Task_2.cpp (hunk 6 above)
Comment (shotvlad) [reply Task_2/Task_2.cpp]: image

File: Task_2/Task_2.cpp (hunk 7 above)
Comment (shotvlad) [reply Task_2/Task_2.cpp]: image

File: Task_1/Task_1.cpp (hunk 5 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp]: image

File: Task_1/Task_1.cpp (hunk 1 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp]: image

File: Task_1/Task_1.cpp (hunk 2 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp]: image

File: Task_1/Task_1.cpp [hunk 8]
```diff
@@ -1,6 +1,106 @@
...
+}
+
+bool isCheckData(int number, int min, int max)
+{
+    if (min == 1 and number == 0) {}
+    else if (number >= min and number <= max)
+    {
+        return false;
+    }
+    else {}
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp]: It looks weird
Don't understand

File: Task_1/Task_1.cpp [hunk 9]
```diff
@@ -1,6 +1,106 @@
...
+}
+
+bool isCheckDay(int d, int m, int y)
+{
+    if (m == 2 and (d < 1 or d > 28 + (((y % 4) == 0) && (((y % 100) != 0) || (y % 400)) ? 1 : 0)))
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp]: Unreadable code

File: Task_1/Task_1.cpp [hunk 10]
```diff
@@ -1,6 +1,106 @@
...
+    {
+        return false;
+    }
+    else if ((m == 1, 3, 5, 7, 8, 10, 12) and (d < 1 or d > 31))
```
Comment (VadzimAlkhimionak) [inline Task_1/Task_1.cpp]: Unreadable code

File: Task_2/Task_2.cpp [hunk 11]
```diff
@@ -1,6 +1,50 @@
...
+    }
+
+    if (number > 0)
+    {
+        return flip;
+    }
+    else
+    {
+        return -flip;
+    }
```
Comment (VadzimAlkhimionak) [inline Task_2/Task_2.cpp]: for what?

**Review by VadzimAlkhimionak — COMMENTED** [review]
Some comments

File: Task_2/Task_2.cpp (hunk 11 above)
Comment (shotvlad) [reply Task_2/Task_2.cpp:49]: Если число отрицательное

File: Task_1/Task_1.cpp (hunk 9 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp]: (m == 2 and (d < 1 or d > 28 + ((y % 4 == 0) and ((y % 100 != 0) or (y % 400 == 0)) ? 1 : 0)))
This is a leap year check. 
The task says the date of birth. February 30 doesn't exist. And on February 29, every 4 years. If the year is a multiple of 100, then it is not a leap year, except for a multiple of 400.

File: Task_2/Task_2.cpp (hunk 11 above)
Comment (shotvlad) [reply Task_2/Task_2.cpp]: If the number is negative, then it will remain negative.

File: Task_1/Task_1.cpp (hunk 8 above)
Comment (shotvlad) [reply Task_1/Task_1.cpp:70]: bool isCheckData(int number, int min, int max)
{
    if (number < min or number > max)
    {
        return true;
    }
    else 
    { 
        return false;
    }
}

Comment (shotvlad) [reply Task_1/Task_1.cpp]: image

**Pull request merged**